src/shell/layer/mod.rs

```diff
@@ -0,0 +1,489 @@
+    pub fn map<D>(
+        self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+        layer_state: &mut LayerState,
+        surface: wl_surface::WlSurface,
+        zdepth: ZDepth,
+    ) -> Result<Layer, LayerSurfaceError>
+    where
+        D: Dispatch<zwlr_layer_surface_v1::ZwlrLayerSurfaceV1, UserData = LayerData> + 'static,
+    {
+        // The z-depth is required in ext-layer-shell-v1 but is not part of the factory request. So the param
+        // will stay for ext-layer-shell-v1 support.
+
+        let layer_shell =
+            layer_state.wlr_layer_shell().ok_or(LayerSurfaceError::MissingRequiredGlobals)?;
+        let layer_surface = layer_shell.get_layer_surface(
+            conn,
+            &surface,
+            self.output.as_ref(),
+            zdepth.into(),
+            self.namespace.unwrap_or_default(),
+            qh,
+            LayerData {},
+        )?;
+
+        // Set data for initial commit
+        if let Some(size) = self.size {
+            layer_surface.set_size(conn, size.0, size.1);
+        }
+
+        if let Some(anchor) = self.anchor {
+            // We currently rely on the bitsets matching
+            layer_surface
+                .set_anchor(conn, zwlr_layer_surface_v1::Anchor::from_bits_truncate(anchor.bits()));
+        }
+
+        if let Some(zone) = self.zone {
+            layer_surface.set_exclusive_zone(conn, zone);
+        }
+
+        if let Some(margin) = self.margin {
+            layer_surface.set_margin(conn, margin.0, margin.1, margin.2, margin.3);
+        }
+
+        if let Some(interactivity) = self.interactivity {
+            layer_surface.set_keyboard_interactivity(conn, interactivity.into())
+        }
+
+        // Initial commit
+        surface.commit(conn);
+
+        Ok(Layer {
+            kind: LayerKind::Wlr(layer_surface),
+            wl_surface: surface,
+            primary: true,
+            death_signal: Arc::new(AtomicBool::new(false)),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Layer {
+    kind: LayerKind,
+
+    wl_surface: wl_surface::WlSurface,
+
+    /// Whether this is the primary handle to the layer.
+    ///
+    /// This is only true for [`Layer`] given the user from [`LayerBuilder::map`]. Since we pass
+    /// a reference to a [`Layer`] in some traits the user implements, we need to make sure the layer isn't
+    /// actually destroyed while the user still holds the layer. If this field is true, the drop implementation
+    /// will mark the layer as dead and will clean up when possible.
+    pub(crate) primary: bool,
+
+    /// Indicates whether the primary handle to the layer has been destroyed.
+    ///
+    /// Since we can't destroy wayland objects without a connection handle, we need to mark the layer for
+    /// cleanup.
+    pub(crate) death_signal: Arc<AtomicBool>,
+}
+
+impl PartialEq for Layer {
+    fn eq(&self, other: &Self) -> bool {
+        self.kind == other.kind
+    }
+}
+
+impl Layer {
```

Comment:
This represents a layer surface, not a layer. Let's name it `LayerSurface` instead (and similarly for the related types like the builder).